src/hotspot/share/compiler/compilerOracle.cpp

```diff
@@ -634,20 +634,35 @@ void skip_comma(char* &line) {
   }
 }
 
-static bool parseEnumValueAsUintx(enum CompileCommand option, const char* line, uintx& value, int& bytes_read, char* errorbuf, const int buf_size) {
+// Parse an uintx-based option value. Also takes care of parsing enum values for options that are enums.
+// Returns true if ok, false if the value could not be parsed.
+static bool parseUintxValue(enum CompileCommand option, const char* line, uintx& value, int& bytes_read) {
+
+#define IF_ENUM_STRING(S, CMD)                \
+  if (strncasecmp(line, S, strlen(S)) == 0) { \
+    bytes_read += (int)strlen(S);             \
+    CMD                                       \
+    return true;                              \
+  }
+  // Parse MemStat as enum
   if (option == CompileCommand::MemStat) {
-    if (strncasecmp(line, "collect", 7) == 0) {
-      value = (uintx)MemStatAction::collect;
-    } else if (strncasecmp(line, "print", 5) == 0) {
-      value = (uintx)MemStatAction::print;
-      print_final_memstat_report = true;
-    } else {
-      jio_snprintf(errorbuf, buf_size, "MemStat: invalid value expected 'collect' or 'print' (omitting value means 'collect')");
-    }
-    return true; // handled
+    IF_ENUM_STRING("collect", {
+        value = (uintx)MemStatAction::collect;
+    });
+    IF_ENUM_STRING("print", {
+        value = (uintx)MemStatAction::print;
+        print_final_memstat_report = true;
```

Comment:
Ouch, I just realized that we can't differentiate between being provided with the literal number 2 and `MemStatAction::print` anymore since you moved the literal number parsing into this function. That means that we can't set `print_final_memstat_report` after returning from this function.

Reply:
> Ouch, I just realized that we can't differentiate between being provided with the literal number 2 and `MemStatAction::print` anymore since you moved the literal number parsing into this function. That means that we can't set `print_final_memstat_report` after returning from this function.

Yes, its a big annoyance with unnecessary  templatization in this file. The natural point of "print_final.." would be in register_command, but I cannot (easily) query values there because value is T. So I cannot just do a "if type is uintx and option is Memstat and value is print then do xxx".

This code would gain IMHO by being dumbed down. A multi-type value can just as well be expressed with a union, no templates needed. Especially since we keep the type info around at runtime anyway.

Reply:
>This code would gain IMHO by being dumbed down

Agreed, seems like it's had a lot of "just one more feature" commits over time.